In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# 데이터 불러오기
train = pd.read_csv('/kaggle/input/moneyball/train.csv')
test = pd.read_csv('/kaggle/input/moneyball/test.csv')

# 데이터 구조 확인
print(train.head())
print(train.info())

df = train.copy()

   ID   W    G    R    AB     H    1B   2B  3B   HR  ...  HBP  BBHBP  SF  \
0   1  60  162  558  5336  1280   891  247  31  111  ...    0    460   0   
1   3  86  161  677  5620  1492  1147  216  46   83  ...    0    531   0   
2   5  79  162  679  5550  1496  1088  262  63   83  ...    0    543   0   
3   7  96  162  780  5536  1458   990  265  40  163  ...    0    561   0   
4   9  96  162  688  5441  1354   905  247  21  181  ...   55    597  35   

   Outs  Outsinplay   RA        BA       OBA       SLG       OPS  
0  4056        3216  740  0.239880  0.300207  0.360195  0.660402  
1  4128        3376  643  0.265480  0.328890  0.364591  0.693480  
2  4054        3175  710  0.269550  0.334646  0.384324  0.718971  
3  4078        3145  682  0.263367  0.331146  0.414017  0.745164  
4  4087        2703  548  0.248851  0.321258  0.401764  0.723022  

[5 rows x 24 columns]
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 581 entries, 0 to 580
Data columns (total 24 columns):
 #   Column  

ID: 팀명

W: (Wins) 한 시즌에 승리한 경기수 (Target Value)

G: (Games) 경기수 (보통 162게임이 기본값; 2018년 대회에서 코로나 단축시즌같은 건 없을 듯)

R: (Runs) 득점

AB: (At Bats)타수

H: (Hits) 안타

1B / 2B / 3B / HR: 단타, 2루타, 3루타, 홈런

BB: (Base on Balls)볼넷

SO: (Strikeouts) 삼진

SB: (Stolen Bases) 도루 성공 횟수

CS: (Caught Stealing) 도루 실패 횟수

HBP: (Hit by Pitch) 몸에 맞는 공

BBHBP: BB + HBP

SF: (Sacrifice Flies) 희생플라이

Outs: 팀이 기록한 아웃카운트 총합

Outsinplay: 인플레이 상황에서 발생한 아웃

RA: (Runs Allowed) 총 실점

BA: (Batting Average) 타율

OBA: (On-Base Average) (= OBP) 출루율

SLG: (Slugging Percentage) 장타율

OPS: OBA + SLG

In [6]:
df.describe()

ID           W           G           R           AB  \
count   581.000000  581.000000  581.000000  581.000000   581.000000   
mean    581.000000   79.413081  158.969019  713.099828  5432.340792   
std     335.729057   12.536626   11.372186  100.936394   396.338140   
min       1.000000   37.000000  103.000000  329.000000  3493.000000   
25%     291.000000   71.000000  162.000000  647.000000  5462.000000   
50%     581.000000   80.000000  162.000000  712.000000  5518.000000   
75%     871.000000   89.000000  162.000000  783.000000  5583.000000   
max    1161.000000  116.000000  164.000000  993.000000  5781.000000   

                 H           1B          2B          3B          HR  ...  \
count   581.000000   581.000000  581.000000  581.000000  581.000000  ...   
mean   1419.020654   976.506024  262.604131   32.425129  147.485370  ...   
std     130.178377    93.027538   42.435244   10.146172   40.995229  ...   
min     797.000000   576.000000  119.000000   11.000000   49.000000  ...   
25%    1369.000000   937.000000  237.000000   25.000000  116.000000  ...   
50%    1432.000000   989.000000  267.000000   31.000000  147.000000  ...   
75%    1499.000000  1037.000000  289.000000   38.000000  176.000000  ...   
max    1684.000000  1193.000000  376.000000   67.000000  257.000000  ...   

              HBP       BBHBP          SF         Outs   Outsinplay  \
count  581.000000  581.000000  581.000000   581.000000   581.000000   
mean    21.846816  544.079174   17.323580  4013.320138  3043.872633   
std     28.656809   82.461301   22.231198   290.986155   253.618335   
min      0.000000  278.000000    0.000000  2601.000000  2022.000000   
25%      0.000000  495.000000    0.000000  4044.000000  2952.000000   
50%      0.000000  544.000000    0.000000  4085.000000  3072.000000   
75%     50.000000  597.000000   41.000000  4115.000000  3207.000000   
max    103.000000  802.000000   70.000000  4212.000000  3447.000000   

                RA          BA         OBA         SLG         OPS  
count   581.000000  581.000000  581.000000  581.000000  581.000000  
mean    713.444062    0.261091    0.327273    0.402523    0.729796  
std     104.803539    0.012078    0.014118    0.032048    0.043866  
min     343.000000    0.226356    0.284100    0.326515    0.614404  
25%     651.000000    0.252307    0.317614    0.378841    0.699169  
50%     710.000000    0.261026    0.326602    0.401113    0.728662  
75%     779.000000    0.269683    0.336702    0.425832    0.758919  
max    1103.000000    0.292515    0.363665    0.483839    0.845694  

[8 rows x 24 columns]

**2018년 대회 & 데이터: 대회 시점 기준 40년 이상의 MLB 팀 기록**

- 경기수: 대부분 162게임이지만, 103게임, 164게임 등 이상치 존재 (노조 파업 or 162게임 체제 이전 or 타이브레이커)
- HBP/ SF: 상당수 데이터가 0 (말이 안됨) => 데이터 오류 or 집계하지 않았던 시기
  * Domain Knowledge: MLB에서 희생플라이는 1954년부터 공식적으로 집계됨.

In [7]:
# 1. SF(희생플라이)가 0인 데이터의 특성 파악 (과거 데이터인지 판별)
zero_sf_df = df[df['SF'] == 0]
normal_sf_df = df[df['SF'] > 0]

print(f"SF가 0인 데이터 개수: {len(zero_sf_df)}")
print(f"SF가 0이 아닌 데이터 개수: {len(normal_sf_df)}")

# 두 그룹의 홈런(HR)과 삼진(SO) 평균 비교
print("\n[비교] SF 유무에 따른 평균 기록:")
comparison = pd.DataFrame({
    'Metric': ['Mean HR', 'Mean SO', 'Mean G'],
    'SF=0 Group': [zero_sf_df['HR'].mean(), zero_sf_df['SO'].mean(), zero_sf_df['G'].mean()],
    'SF>0 Group': [normal_sf_df['HR'].mean(), normal_sf_df['SO'].mean(), normal_sf_df['G'].mean()]
})
print(comparison)

# Insight: 만약 SF=0 그룹의 HR/SO가 현저히 낮다면, 이는 '옛날 야구' 데이터이므로 학습에서 제외하는 것을 추천.

SF가 0인 데이터 개수: 354
SF가 0이 아닌 데이터 개수: 227

[비교] SF 유무에 따른 평균 기록:
    Metric  SF=0 Group   SF>0 Group
0  Mean HR  134.596045   167.585903
1  Mean SO  875.025424  1116.696035
2   Mean G  157.059322   161.947137


In [8]:
# 2. 파생 변수 검증 (OPS 계산이 맞는지 확인)
# 데이터에 있는 OPS 값과 직접 계산한 OPS 값의 차이(Error) 확인
# OBA + SLG = OPS
df['Calculated_OPS'] = df['OBA'] + df['SLG']
df['OPS_Error'] = df['OPS'] - df['Calculated_OPS']

print("\n[검증] 제공된 OPS와 계산된 OPS의 차이:")
print(df['OPS_Error'].describe())

# 만약 차이가 0.001 이상 나는 데이터가 많다면, 제공된 OPS 대신 직접 계산한 값을 써야 함.


[검증] 제공된 OPS와 계산된 OPS의 차이:
count    5.810000e+02
mean     5.163514e-12
std      5.068233e-10
min     -1.000000e-09
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      1.000000e-09
Name: OPS_Error, dtype: float64


In [ ]:
# 3. 승률과 득실점의 상관관계 (피타고리안 기대 승률 검증)
# 경기수(G)가 다르므로 승수(W) 대신 승률(Win_Rate) 사용
df['Win_Rate'] = df['W'] / df['G']

# 득실차(Run Differential) 계산
df['Run_Diff'] = df['R'] - df['RA']

# 시각화
plt.figure(figsize=(10, 6))
sns.scatterplot(x='Run_Diff', y='Win_Rate', data=df, alpha=0.6)
plt.title('Run Differential vs Win Rate') # 득실차와 승률의 관계
plt.xlabel('Run Differential (R - RA)')
plt.ylabel('Win Rate')
plt.axvline(0, color='red', linestyle='--') # 0점 기준선
plt.axhline(0.5, color='red', linestyle='--') # 5할 승률 기준선
plt.grid(True)
plt.show()